src/lib/mina_base/zkapp_command.ml

```diff
@@ -1184,40 +1238,123 @@ end = struct
-                (p, None) ) )
+              match p.body.authorization_kind with
+              | Proof vk_hash ->
+                  let prioritized_vk =
+                    (* only lookup _past_ vk setting, ie exclude the new one we
+                     * potentially set in this account_update (use the non-'
+                     * vks_overrided) . *)
+                    match Account_id.Map.find !vks_overridden account_id with
+                    | Some (Some vk) -> (
+                        match
+                          ok_if_vk_hash_expected ~got:vk ~expected:vk_hash
+                        with
+                        | Ok vk ->
+                            vk
+                        | Error err ->
+                            return (Error err) )
+                    | Some None ->
+                        (* we explicitly have erased the key *)
+                        let err =
+                          Error.create
+                            "No verification key found for proved account \
+                             update: the verification key was removed by a \
+                             previous account update"
+                            ("account_id", account_id)
+                            [%sexp_of: string * Account_id.t]
+                        in
+                        return (Error err)
+                    | None -> (
+                        (* we haven't set anything; lookup the vk in the fallback *)
+                        match find_vk vk_hash account_id with
+                        | Error e ->
+                            return (Error e)
+                        | Ok vk ->
+                            vk )
+                  in
+                  Account_id.Table.update tbl account_id ~f:(fun _ ->
+                      With_hash.hash prioritized_vk ) ;
+                  (* return the updated overrides *)
+                  vks_overridden := vks_overriden' ;
+                  (p, Some prioritized_vk)
+              | _ ->
+                  vks_overridden := vks_overriden' ;
+                  (p, None) )
         in
         Ok { fee_payer; account_updates; memo } )
+
+  module Map_cache = struct
+    type 'a t = 'a Zkapp_basic.F_map.Map.t
+
+    let empty = Zkapp_basic.F_map.Map.empty
+
+    let find = Zkapp_basic.F_map.Map.find
+
+    let set = Zkapp_basic.F_map.Map.set
+  end
+
+  module Singleton_cache = struct
+    type 'a t = (Zkapp_basic.F.t * 'a) option
+
+    let empty = None
+
+    let find t key =
+      match t with
+      | None ->
+          None
+      | Some (k, v) ->
+          if Zkapp_basic.F.equal key k then Some v else None
+
+    let set _ ~key ~data = Some (key, data)
+  end
+
+  module Make_create_all (Cache : sig
+    type 'a t
+
+    val empty : 'a t
+
+    val find : 'a t -> Zkapp_basic.F.t -> 'a option
+
+    val set : 'a t -> key:Zkapp_basic.F.t -> data:'a -> 'a t
+  end) =
+  struct
+    let create_all (cmds : T.t list)
+        ~(find_vk :
```

Comment:
Calling `find_vk` like this one at a time is going to be inefficient, since each invocation will fallback to the db ledger if the vk was not modified in any of the frontier masks. I fixed a number of issues on mainnet this year related to not batching db operations, and code like this could reintroduce similar performance issues if we need to lookup a lot of vks. We should consider fixing this now by just rewriting `find_vk` to work on batches.

Reply:
I've decided to implement this in a separate PR. The work for this is already tracked in this prior issue https://github.com/MinaProtocol/mina/issues/11705.